rfc/rfc-0000-k8s-events.md

```diff
@@ -0,0 +1,176 @@
+| StampedObjectObjectGetError | `<resource name>` could not retrieve `<group>.<kind>` due to error: `<error message>` | Loading a resources is failing due to a client.get issue (missing is not an error) - this could be spammy | Blueprint |
+| StampedObjectGarbageCollected | `<group>.<kind>/<name>` is no longer referenced | This owner has selected a different template, either by supply chain selection or templating, and this object is no longer needed | Owner |
+| ImmutableStampedObjectGarbageCollected | `<n> * <group>.<kind>` historical objects deleted due to garbage collection policy | This runnable's GC policy has caused `n` objects to be removed | Runnable |
+| StampedObjectInvalid | `<resource name>` could not be applied as `<group>.<kind>/<name>` due to API server error `<error message>` | This object was (probably) malformed. | Owner |
+| StampedObjectInvalid | `<resource name>` could not be applied as `<group>.<kind>` due to API server error `<error message>` | This object was (probably) malformed. This lets operators know their templates might have issues | Blueprint |
+| StampedObjectKindNotFound | `<resource name>` could not be applied because `<group>.<kind>/<name>` does not exist on this cluster | Did someone forget to install the CRDs? Otherwise it's a malformed template | Owner |
+| StampedObjectKindNotFound | `<resource name>` could not be applied because `<group>.<kind>` does not exist on this cluster | Did someone forget to install the CRDs? Otherwise it's a malformed template | Blueprint |
+| StampedObjectApplied | `<resource name>` was applied as `<group>.<kind>/<name>` | a stamped object needed to be created/updated | Owner |
+| StampedObjectKindChanged | `<resource name>` was `<old group>.<old kind>`, now `<new group>.<new kind>` | YTT selection for a template GVK changed, or templated values in GVK changed | Owner |
+| ResourceOutputChanged | `<resource name>` found a new output in `<group>.<kind>/<name>` | a resource produced a new output | Owner |
+| ResourceHealthyStatusChanged | `<resource name>` found a new status in `<group>.<kind>/<name>` | a resource produced a new healthy status | Owner |
+| SupplyChainChanged | supply chain changed from `<old supply chain name>` to `<new supply chain name>` | Workload selected for a new or different supply chain. Note: `none` is a possible name. | Owner |
+| WorkloadSelected | supply chain matched a workload | A workload was selected for (really, this is when the supplyChainRef changes). This makes the count a useful metric | Blueprint |
+
+* Cache misses? (Not sure about this, we'll miss if the controller restarts or with HA Elections (future concern))
+
+**Note:** Blueprint event `messages` are usually designed to aggregate on the resource kind, not the individual
+resource (we omit `<resource name>`)
```

Comment:
| The owner information (a namespaces resource) likely shouldn't be recorded against the blueprint resource (a cluster scoped resource). 

Agreed, so we do not.

The User Journey I have in mind when adding these:

* I am an operator.
* A developer approaches me with an issue in their workload
* I see the issue, and it can be fixed by changing some configuration
* I wonder who else is affected by this kind of issue, and look to the supply chain for 'all workload' aggregates of those kinds of events.

I am in no rush to introduce the blueprint events, so could put them in an addenda for now, as a possibility for Operator facing UX in the future.

Reply:
Oh I lie, we do name the resource. An oversight on my behalf, I don't want events stacking up for each resource instance. Will fix and commit some changes.

Reply:
FYI this is an overload issue. `<resource name>` is not an instance name (`metadata.name`) but the name in the supply chain (`spec.resources[].name`)

Reply:
> _I am in no rush to introduce the blueprint events,_

I'd +1 reducing the scope of this RFC to just owner-related events. wdyt, @sclevine @rawlingsj ?

Reply:
The scope of this RFC is to introduce events and to establish some guidelines. Not to force a list of events.

OTOH, I really believe in the value of bluepring scoped events, _especially_ the BlueprintChanged event.